### PR TITLE
overlord/snapstate, overlord/devicestate: exclusive change conflict check

### PR DESCRIPTION
**Pull request by bboozzoo**
Add a helper for checking for conflicts for changes that are the only ones that run at any given time. Add a conflict check for create-recovery-system change introduced in #10254

### CODE REVIEW

File: overlord/devicestate/devicestate_remodel_test.go [hunk 1]
```diff
@@ -909,13 +909,24 @@ func (s *deviceMgrRemodelSuite) TestReregRemodelClashAnyChange(c *C) {
 		"required-snaps": []interface{}{"new-required-snap-1", "new-required-snap-2"},
 		"revision":       "1",
 	})
+	s.newFakeStore = func(devBE storecontext.DeviceBackend) snapstate.StoreService {
+		mod, err := devBE.Model()
+		c.Check(err, IsNil)
+		if err == nil {
```
Comment (mardy) [inline overlord/devicestate/devicestate_remodel_test.go:915]: doen't the check on the line above ensure that this condition will be hit?

Comment (bboozzoo) [reply overlord/devicestate/devicestate_remodel_test.go:915]: No, c.Check() does not stop a test execution if the condition is false, that would be c.Assert(). OTOH, calling an assert in callback that can run under taskrunner (i.e. in a separate goroutine and cause a panic there), is problematic and often makes the failing test either hang or produce a backtrace that obscures the original failure.

Also, on a closed look, with the changed conflict check, the callback would not be called at all. I'll push a patch

File: overlord/snapstate/conflict.go [hunk 2]
```diff
@@ -90,6 +90,34 @@ func affectedSnaps(t *state.Task) ([]string, error) {
 	return nil, nil
 }
 
+// CheckChangeKindExclusiveConflict checks for conflicts with a new change which
+// must be run when no other changes are running.
+func CheckChangeKindExclusiveConflict(st *state.State, newChangeKind string) error {
```
Comment (stolowski) [inline overlord/snapstate/conflict.go:95]: The name sounds a bit unclear, we want to conflict with any non-ready change regardless of its kind,  maybe `CheckChangeConflictAnyRunningChange`? The comment should say that newChangeKind is only for the error message and doesn't affect conflict logic, e.g.
```
CheckChangeConflictAnyRunningChange checks that there are no unready changes and if so, returns ChangeConflictError. newChangeKind is only used for the error message.
```

File: overlord/snapstate/conflict.go [hunk 3]
```diff
@@ -90,6 +90,34 @@ func affectedSnaps(t *state.Task) ([]string, error) {
...
+	for _, chg := range st.Changes() {
```
[the remark this answers is not in the document]
Comment (bboozzoo) [reply overlord/snapstate/conflict.go:94]: refactored and renamed

File: overlord/devicestate/devicestate.go [hunk 4]
```diff
@@ -555,7 +555,7 @@ func Remodel(st *state.State, new *asserts.Model) (*state.Change, error) {
 
 	// Do we do this only for the more complicated cases (anything
 	// more than adding required-snaps really)?
-	if err := snapstate.CheckChangeKindExclusiveConflict(st, "remodel"); err != nil {
+	if err := snapstate.CheckChangeConflictRunExclusively(st, "remodel"); err != nil {
```
Comment (pedronis) [inline overlord/devicestate/devicestate.go:558]: this reads decently I think

**Review by pedronis — APPROVED** [review]
thank you

**Review by mvo5 — APPROVED** [review]
Thank you

File: overlord/snapstate/conflict.go [hunk 5]
```diff
@@ -90,6 +90,60 @@ func affectedSnaps(t *state.Task) ([]string, error) {
...
+			}
+			return &ChangeConflictError{
+				Message:    "creating recovery system in progress, no other changes allowed until this is done",
+				ChangeKind: "create-recovery-system",
+			}
+		default:
+			if newExclusiveChangeKind != "" {
+				// we want to run a new exclusive change, but other
+				// changes are in progress already
+				msg := fmt.Sprintf("other changes in progress, change %q not allowed until they are done (conflicting change %q)",
```
Comment (pedronis) [inline overlord/snapstate/conflict.go:129]: the parentheses could also go before the ",", it's slightly more cohesive that way but no super strong opinion though

**Review by pedronis — APPROVED** [review]
thanks, small consideration

**Pull request merged**